File: 2019/1/day1.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func main() {
	inFile, _ := ioutil.ReadFile("input")
	lines := strings.Split(string(inFile), "\n")

	fmt.Printf("Part 1: %d\n", part1(lines))
	fmt.Printf("Part 2: %d\n", part2(lines))
}

func part1(lines []string) int {
	fuel := 0
	for _, line := range lines {
		mass, _ := strconv.Atoi(line)
		fuel += cost(mass)
	}
	return fuel
}

func part2(lines []string) int {
	fuel := 0
	for _, line := range lines {
		mass, _ := strconv.Atoi(line)
		fuel += costRec(mass, 0)
	}
	return fuel
}

func cost(mass int) int {
	return mass/3 - 2
}

func costRec(mass int, acc int) int {
	c := cost(mass)
	if c <= 0 {
		return acc
	}
	return costRec(c, acc+c)
}

File: 2019/1/day1.c
#include <stdio.h>
#include <stdlib.h>

int cost(int mass) {
    return mass / 3 - 2;
}

int costRec(int mass, int acc) {
    int c = cost(mass);
    if (c <= 0) {
        return acc;
    }
    return costRec(c, acc+c);
}

int part1() {
    FILE* inputFile = fopen("input", "r");
    char line [10];

    int fuel = 0;
    while (fgets(line, sizeof(line), inputFile)) {
        fuel += cost(atoi(line));
    }
    fclose(inputFile);
    
    return fuel;
}

int part2() {
    FILE* inputFile = fopen("input", "r");
    char line [10];

    int fuel = 0;
    while (fgets(line, sizeof(line), inputFile)) {
        fuel += costRec(atoi(line), 0);
    }
    fclose(inputFile);
    
    return fuel;
}

int main(int argc, char *argv[]) {

    printf("Part 1: %d\n", part1());
    printf("Part 2: %d\n", part2());
    return 0;
}
